### PR TITLE
Add a Bootstrap layout to the CRUD add-on

### LINKED ISSUE #483: Add a Bootstrap layout to the CRUD add-on
## Why

*This is the first brick to the CRUD addon.*

Having a ready-to-use Bootstrap layout, aligned with our design guidelines will help us save time when starting a new 'standard' project.

- Add the CRUD addon option and base file structure
- Add the layout files for Slim ~or ERB~ (confirmed in https://github.com/nimblehq/rails-templates/discussions/484)

## Who Benefits?

Developers on standard admin projects.

### PR DESCRIPTION
**Pull request by malparty**
- close #483

## What happened 👀

- [x] Add the first version of the CRUD addon with a simple bootstrap layout
- [x] Add the ruby script to copy SCSS styles and insert the `import` lines in the root application.scss file
- [x] Add tests

## Insight 📝

> [!IMPORTANT]
> The purpose of this PR is to get a starting point, not to get the perfect layout 👍 
>
> The `look and feel` can be discussed in future issues, I also asked UX team as they have a Figma project for default Rails/Bootstrap apps, but they don't have recommendations for `organism components` (e.g. full menu, sidebar, etc...).
> => I decided to use the Bootstrap examples from the official doc, so I don't need to think too much about it.

SCSS classes (instead of Bootstrap utility classes) will be implemented in this issue:
- #491

## Proof Of Work 📹

image

The Top navbar is responsive, not the sidebar (yet?):

image

### CODE REVIEW

File: .template/addons/crud/app/assets/stylesheets/layouts/_application.scss
Comment (malparty) [inline .template/addons/crud/app/assets/stylesheets/layouts/_application.scss:1]: This file will be reviewed and improved in #491

File: .template/addons/crud/app/assets/stylesheets/layouts/_application.scss [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    padding: rem(16px);
+    margin: rem(16px);
+  }
+
+  &__footer {
+    grid-area: footer;
+    padding: rem(8px);
+  }
+
+  &__nav_bar_top {
```
Comment (sanG-github) [inline .template/addons/crud/app/assets/stylesheets/layouts/_application.scss:29]: I think we should separate them by the hyphen (-) instead
```suggestion
  &__nav-bar-top {
```

File: .template/addons/crud/app/assets/stylesheets/layouts/_application.scss [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      color: var(--bs-emphasis-color);
+      background-color: transparent;
+    }
+    .btn-toggle:hover,
+    .btn-toggle:focus {
+      color: rgba(var(--bs-emphasis-color-rgb), 0.85);
+      background-color: var(--bs-tertiary-bg);
+    }
+
+    .btn-toggle::before {
```
Comment (sanG-github) [inline .template/addons/crud/app/assets/stylesheets/layouts/_application.scss:59]: IMO, I think we can put all the states like `:hover`, `:focus`, or `::before` pseudo-element, ... inside the 
`.btn-toggle` selector.

File: .template/addons/crud/app/assets/stylesheets/layouts/index.scss [hunk 3]
```diff
@@ -0,0 +1 @@
+@import "./application";
```
Comment (sanG-github) [inline .template/addons/crud/app/assets/stylesheets/layouts/index.scss:1]: We can do like this, right? 😛 
```suggestion
@import "application";
```

File: .template/addons/crud/app/views/layouts/application.html.slim.tt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  head
+    meta charset="utf-8"
+    meta name="viewport" content="width=device-width, initial-scale=1.0"
+    title= content_for?(:title) ? yield(:title) : '<%= app_name %>'
+    = csrf_meta_tags
+    = csp_meta_tag
+    = stylesheet_link_tag 'application', :media => 'all'
+    = javascript_include_tag 'application'
+
+  body class=class_names(controller_name.dasherize, action_name.dasherize)
```
Comment (sanG-github) [inline .template/addons/crud/app/views/layouts/application.html.slim.tt:12]: I'm unsure if we should need to keep these classes. 🤔 
```suggestion
  body
```

File: .template/addons/crud/app/views/layouts/application/_nav_bar_top.html.slim [hunk 5]
```diff
@@ -0,0 +1,33 @@
+.app__nav_bar_top
+  nav.navbar.navbar-expand-lg.navbar-light.bg-light
+    .container-fluid
+      a.navbar-brand[href="#"]
+        | NavbarTop
```
Comment (sanG-github) [inline .template/addons/crud/app/views/layouts/application/_nav_bar_top.html.slim:5]: Will it still work without this notation `|`? 💭 
```suggestion
        NavbarTop
```

File: .template/addons/crud/template.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+remove_file 'app/views/layouts/application.html.erb'
+directory 'app/views/layouts/application'
+template 'app/views/layouts/application.html.slim.tt'
+
+# SCSS Layout
+directory 'app/assets/stylesheets/layouts'
+insert_into_file 'app/assets/stylesheets/application.scss', after: %r{// Layouts\n} do
+  <<~SCSS
+    @import './layouts';
```
Comment (sanG-github) [inline .template/addons/crud/template.rb:13]: The same suggestion as above here 🙏🏻 
```suggestion
    @import 'layouts';
```

File: .template/variants/web/app/template.rb [hunk 7]
```diff
@@ -3,13 +3,20 @@
...
+
+if erb_layout_exists
+  insert_into_file erb_layout_file, before: %r{</head>} do
     <<~ERB.indent(2)
       <%= javascript_include_tag "application", "data-turbo-track": "reload", defer: true %>
     ERB
   end
-else
+# The slim layout (CRUD addon) already has the javascript include tag
+elsif !slim_layout_exists
```
Comment (sanG-github) [inline .template/variants/web/app/template.rb:19]: If I understand it correctly 💭 

```suggestion
elsif slim_layout_exists
```

**Review by sanG-github — COMMENTED** [review]
I have some comments and small suggestions 🙏🏻

File: .template/addons/crud/app/assets/stylesheets/layouts/_application.scss (hunk 2 above)
Comment (malparty) [reply .template/addons/crud/app/assets/stylesheets/layouts/_application.scss]: 3270c1d :white_check_mark:

File: .template/addons/crud/app/assets/stylesheets/layouts/_application.scss (hunk 1 above)
Comment (malparty) [reply .template/addons/crud/app/assets/stylesheets/layouts/_application.scss]: 3270c1d :white_check_mark:

File: .template/addons/crud/app/views/layouts/application/_nav_bar_top.html.slim (hunk 5 above)
Comment (malparty) [reply .template/addons/crud/app/views/layouts/application/_nav_bar_top.html.slim:5]: This is using the SLIM template engine. Just adding test means the `<navbartop>` HTML element:

image


```suggestion
        div NavbarTop
```

An alternative would be to be explicit about `div`, but I find the `|` more elegant 🙈

File: .template/addons/crud/app/views/layouts/application.html.slim.tt (hunk 4 above)
Comment (malparty) [reply .template/addons/crud/app/views/layouts/application.html.slim.tt:12]: These classes are actually quite usefull when you need a screen CSS class 💡 

It's added with this in mind: https://nimblehq.co/compass/development/code-conventions/css/#screens

File: .template/variants/web/app/template.rb (hunk 7 above)
Comment (malparty) [reply .template/variants/web/app/template.rb]: This one was confusing. It meant that we throw the error if the layout is neither ERB neither Slim.

Renamed in 0243e3d, let me know if that's clearer or not 😇

**Review by sanG-github — APPROVED** [review]

**Review by khangbui22 — APPROVED** [review]
LGTM! 🚀

**Pull request merged**